flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceSplitEnumerator.java

```diff
@@ -0,0 +1,390 @@
+                        for (int pendingReaderSubtaskId : pendingReaders) {
+                            context.sendEventToSourceReader(
+                                    pendingReaderSubtaskId,
+                                    new SwitchSourceEvent(currentSourceIndex));
+                        }
+                        for (int pendingReaderSubtaskId : pendingReaders) {
+                            currentEnumerator.addReader(pendingReaderSubtaskId);
+                        }
+                        pendingReaders.clear();
+                    }
+                } else {
+                    // source already switched
+                    LOG.debug("Ignoring out of order source event {}", srfe);
+                }
+                return;
+            }
+            this.assignments.remove(subtaskId);
+            LOG.info(
+                    "Reader finished for subtask {} remaining assignments {}",
+                    subtaskId,
+                    assignments);
+            if (this.assignments.isEmpty()) {
+                LOG.debug("No assignments remaining, ready to switch source!");
+                if (currentSourceIndex + 1 < sourceChain.sources.size()) {
+                    switchEnumerator();
+                    // switch all readers prior to sending split assignments
+                    for (int i = 0; i < context.currentParallelism(); i++) {
+                        context.sendEventToSourceReader(
+                                i, new SwitchSourceEvent(currentSourceIndex));
+                    }
+                    // trigger split assignment,
+                    // (initially happens as part of subtask/reader registration)
+                    for (int i = 0; i < context.currentParallelism(); i++) {
+                        LOG.debug("adding reader subtask={} sourceIndex={}", i, currentSourceIndex);
+                        currentEnumerator.addReader(i);
+                    }
+                }
+            }
+        } else {
+            currentEnumerator.handleSourceEvent(subtaskId, sourceEvent);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        currentEnumerator.close();
+    }
+
+    private void switchEnumerator() {
+
+        Object enumeratorState = null;
+        if (currentEnumerator != null) {
+            try {
+                enumeratorState = currentEnumerator.snapshotState(-1);
+                currentEnumerator.close();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+            currentEnumerator = null;
+            currentSourceIndex++;
+        }
+
+        SplitEnumeratorContextProxy delegatingContext =
+                new SplitEnumeratorContextProxy(currentSourceIndex, context, assignments);
+        Source<?, ? extends SourceSplit, Object> source =
+                (Source) sourceChain.sources.get(currentSourceIndex).f0;
+        HybridSource.CheckpointConverter<Object, Object> converter =
```

Comment:
`HybridSource.CheckpointConverter<Object, Object>` could be replaced with `HybridSource.CheckpointConverter<?, ?> converter `.